python/tskit/metadata.py

```diff
@@ -120,6 +128,310 @@ def decode(self, data: bytes) -> bytes:
+            if do_sort:
+                # Python sort is stable so we can do the sorts in reverse priority
+                items = sorted(items, key=lambda k_v: k_v[0])
+                items = sorted(items, key=lambda k_v: k_v[1].get("index", 0))
+            items = ((k, cls.order_by_index(v, k == "properties")) for k, v in items)
+            if do_sort:
+                return collections.OrderedDict(items)
+            else:
+                return dict(items)
+        elif isinstance(obj, list) or isinstance(obj, tuple):
+            return [cls.order_by_index(v, False) for v in obj]
+        else:
+            return obj
+
+    @classmethod
+    def make_decode(cls, sub_schema):
+        """
+        Create a function that can decode objects of this schema
+        """
+        return {
+            "array": StructCodec.make_array_decode,
+            "object": StructCodec.make_object_decode,
+            "string": StructCodec.make_string_decode,
+            "null": StructCodec.make_null_decode,
+            "number": StructCodec.make_numeric_decode,
+            "integer": StructCodec.make_numeric_decode,
+            "boolean": StructCodec.make_numeric_decode,
+        }[sub_schema["type"]](sub_schema)
+
+    @classmethod
+    def make_array_decode(cls, sub_schema):
+        element_decoder = StructCodec.make_decode(sub_schema["items"])
+        array_length_f = "<" + sub_schema.get("arrayLengthFormat", "L")
+        array_length_size = struct.calcsize(array_length_f)
+        exhaust_buffer = sub_schema.get("noLengthEncodingExhaustBuffer", False)
+
+        def array_decode(buffer):
+            print(array_length_f)
+            array_length = struct.unpack(
+                array_length_f, bytes(islice(buffer, array_length_size))
+            )[0]
+            return [element_decoder(buffer) for _ in range(array_length)]
+
+        def array_decode_exhaust(buffer):
+            ret = []
+            while True:
+                try:
+                    ret.append(element_decoder(buffer))
+                except struct.error as e:
+                    if "unpack requires a buffer" in str(e):
+                        break
+                    else:
+                        raise e
+            return ret
+
+        if exhaust_buffer:
+            return array_decode_exhaust
+        else:
+            return array_decode
+
+    @classmethod
+    def make_object_decode(cls, sub_schema):
+        sub_decoders = {
+            key: StructCodec.make_decode(prop)
+            for key, prop in sub_schema["properties"].items()
+        }
+        return lambda buffer: {
+            key: sub_decoder(buffer) for key, sub_decoder in sub_decoders.items()
+        }
+
+    @classmethod
+    def make_string_decode(cls, sub_schema):
+        f = "<" + sub_schema["binaryFormat"]
+        size = struct.calcsize(f)
+        encoding = sub_schema.get("stringEncoding", "utf-8")
+        nullTerminated = sub_schema.get("nullTerminated", False)
```

Comment:
nitpick: we're using snake_case for Python variables elsewhere.

Reply:
Heh, too much JSON! Fixed in https://github.com/tskit-dev/tskit/pull/601/commits/f65bc76d411bd231877dc3c8dc45510479e381d7